#!/bin/bash

#########################################################
# 
# Platform: NCI Gadi HPC
# Description: concatenate the HLA gene fastq files output during
# split parallel alignment into one HLA gene per sample
# Usage: first, run 'bash cat_HLA_make_input.sh <cohort_name>', 
# then execute this script with 'qsub rcat_HLA_run_parallel.pbs'
# Details:
#	The modifed BWAkit command used during parallel alignment
#	of split fastq writes a fastq output per HLA gene per sample
#	per fastq input. This means each sample with have thousdands of
#	HLA fastq, some of which will be empty. This set of scripts
#	concatenates them into one HLA gene fastq per sample, so 
#	there will be 6 HLA fastq per sample in 'HLA_fastq_cat'. After 
#	this script has run successfully, the split HLA fastq in 'HLA-fastq' 
#	should be deleted. Allow 1 CPU and 4 GB RAM per sample. Time is 
#	around 6 minutes for 30X and 11 minutes for 60X so while this can 
#	be run as separate tumour/normal jobs to increase CPU efficiency, the 
#	SU saving will be negligible given the low CPU hours used
#
# Author: Cali Willet
# cali.willet@sydney.edu.au
# Date last modified: 31/07/2020
#
# If you use this script towards a publication, please acknowledge the
# Sydney Informatics Hub (or co-authorship, where appropriate).
#
# Suggested acknowledgement:
# The authors acknowledge the scientific and technical assistance 
# <or e.g. bioinformatics assistance of <PERSON>> of Sydney Informatics
# Hub and resources and services from the National Computational 
# Infrastructure (NCI), which is supported by the Australian Government
# with access facilitated by the University of Sydney.
# 
#########################################################

#PBS -P <project>
#PBS -N cat-HLA
#PBS -l walltime=00:20:00
#PBS -l ncpus=22
#PBS -l mem=88GB
#PBS -q normal
#PBS -W umask=022
#PBS -l wd
#PBS -o ./Logs/cat-HLA.o
#PBS -e ./Logs/cat-HLA.e
#PBS -lstorage=scratch/<project>

module load openmpi/4.0.2
module load nci-parallel/1.0.0

set -e

SCRIPT=./cat_HLA.pl  
INPUTS=./Inputs/cat_HLA.inputs  

NCPUS=1 

mkdir -p HLA_fastq_cat


#########################################################
# Do not edit below this line (unless running on a node that 
# does not have 48 CPU, in which case edit the value of 'CPN'
#########################################################

CPN=48 #CPUs per node 
M=$(( CPN / NCPUS )) #tasks per node

sed "s|^|${SCRIPT} |" ${INPUTS} > ${PBS_JOBFS}/input-file

mpirun --np $((M * PBS_NCPUS / CPN)) \
        --map-by node:PE=${NCPUS} \
        nci-parallel \
        --verbose \
        --input-file ${PBS_JOBFS}/input-file
